### PR TITLE
Brave News won't include items with non-http-or-https destination urls

### PR DESCRIPTION
**Pull request by petemill**
Applies to both the combined feed (feed.json hosted on Brave's PCDN) and direct feeds (i.e. RSS feeds)

Resolves https://github.com/brave/brave-browser/issues/27602

## Submitter Checklist:

- [ ] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/brave_today/browser/direct_feed_controller.cc [hunk 1]
```diff
@@ -373,28 +373,8 @@ void DirectFeedController::DownloadFeedContent(const GURL& feed_url,
...
-        std::sort(articles.begin(), articles.end(),
-                  [](mojom::ArticlePtr& a, mojom::ArticlePtr& b) {
-                    return (a.get()->data->score < b.get()->data->score);
-                  });
-        double variety = 2.0;
-        for (auto& entry : articles) {
-          entry->data->score = entry->data->score * variety;
-          variety = variety * 2.0;
-        }
+        DirectFeedController::BuildArticles(articles, response->data,
```
Comment (petemill) [inline components/brave_today/browser/direct_feed_controller.cc:376]: Moved this logic to a testable function

File: components/brave_today/browser/direct_feed_controller.cc [hunk 2]
```diff
@@ -403,6 +383,38 @@ void DirectFeedController::DownloadFeedContent(const GURL& feed_url,
...
+void DirectFeedController::BuildArticles(Articles& articles,
+                                         const FeedData& data,
+                                         const std::string& publisher_id) {
+  for (auto entry : data.items) {
+    auto item = RustFeedItemToArticle(entry);
+    item->data->publisher_id = publisher_id;
+    // Only allow http and https links
+    if (!item->data->url.SchemeIsHTTPOrHTTPS()) {
+      continue;
+    }
```
Comment (petemill) [inline components/brave_today/browser/direct_feed_controller.cc:396]: This bypasses non-http(s) items from direct-downloaded RSS feeds

File: components/brave_today/browser/feed_parsing.cc [hunk 3]
```diff
@@ -61,6 +63,10 @@ bool ParseFeedItem(const base::Value& feed_item_raw,
     VLOG(1) << "Could not parse item url: " << url_raw;
     return false;
   }
+  if (!url.SchemeIsHTTPOrHTTPS()) {
+    VLOG(1) << "Item url was not HTTP or HTTPS: " << url.spec();
+    return false;
+  }
```
Comment (petemill) [inline components/brave_today/browser/feed_parsing.cc:69]: This bypasses non-http(s) items from the combined feed hosted by Brave

**Review by diracdeltas — APPROVED** [review]

**Review by fmarier — APPROVED** [review]

**Review by bridiver — APPROVED** [review]

File: components/brave_today/browser/feed_parsing.cc (hunk 3 above)
Comment (fallaciousreasoning) [reply components/brave_today/browser/feed_parsing.cc:69]: Is there a corresponding backend change? I think ideally we wouldn't be including any non-https items in the combined feed in the first place.

File: components/brave_today/browser/feed_parsing.h [hunk 4]
```diff
@@ -14,6 +14,10 @@
 
 namespace brave_news {
 
+// Convert from the "combined feed" hosted remotely to Brave News mojom items.
+// TODO(petemill): Rename this file to combined_feed_parsing.h or similar,
+// in order to differentiate the "Combined Feed" from
+// a "Direct Feed" (a.k.a RSS).
```
Comment (fallaciousreasoning) [inline components/brave_today/browser/feed_parsing.h:20]: :+1:

**Review by fallaciousreasoning — APPROVED** [review]
LGTM

**Review by sangwoo108 — APPROVED** [review]
LGTM 👍

File: components/brave_today/browser/feed_parsing.cc (hunk 3 above)
Comment (petemill) [reply components/brave_today/browser/feed_parsing.cc:69]: I would hope it wouldn't include it but doesn't hurt to double-check. And there's always the possibility someone points the browser to a third party combined feed source via the cli param...

**Comment by kjozwiak** [comment]
Verification **PASSED** on `Win 11 x64` using the following build(s):

```
Brave | 1.49.11 Chromium: 109.0.5414.80 (Official Build) nightly (64-bit)
-- | --
Revision | 0f69b168d36a06cace4365e9f029fa987afa5633-refs/branch-heads/5414@{#1178}
OS | Windows 11 Version 22H2 (Build 22621.963)
```

Verification notes can be found via https://github.com/brave/internal/issues/987#issuecomment-1376133449.

**Pull request merged**